core/sequencer/metadata/sourceslice.go

```diff
@@ -15,17 +15,54 @@
 package metadata
 
 import (
+	"fmt"
+	"math"
 	"time"
 
 	spb "github.com/google/keytransparency/core/sequencer/sequencer_go_proto"
 )
 
-// Source returns a wrapper for SourceSlice
-func Source(s *spb.MapMetadata_SourceSlice) SourceSlice {
-	return SourceSlice{s: s}
+var minTime = time.Unix(0, math.MinInt64) // 1677-09-21 00:12:43.145224192
+var maxTime = time.Unix(0, math.MaxInt64) // 2262-04-11 23:47:16.854775807
+
+// validateTime determines whether a timestamp is valid.
+// Valid timestamps can be represented by an int64 number of microseconds from the epoch.
+//
+// Every valid timestamp can be represented by a time.Time, but the converse is not true.
+func validateTime(ts time.Time) error {
+	if ts.Before(minTime) {
+		return fmt.Errorf("timestamp %v before %v", ts, minTime)
+	}
+	if ts.After(maxTime) {
+		return fmt.Errorf("timestamp %v after %v", ts, maxTime)
+	}
+	return nil
+}
+
+// New creates a new source slice from time objects.
+// Returns an error if the time objects cannot be represented correctly.
+func New(logID int64, low, high time.Time) (*SourceSlice, error) {
+	if err := validateTime(low); err != nil {
+		return nil, err
+	}
+	if err := validateTime(high); err != nil {
+		return nil, err
+	}
+	return &SourceSlice{s: &spb.MapMetadata_SourceSlice{
+		LogId:            logID,
+		LowestInclusive:  low.UnixNano(),
+		HighestExclusive: high.UnixNano(),
+	}}, nil
+}
+
+// FromProto returns a wrapper for SourceSlice
+func FromProto(s *spb.MapMetadata_SourceSlice) *SourceSlice {
+	return &SourceSlice{s: s}
 }
 
 // SourceSlice defines accessor and conversion methods for MapMetadata_SourceSlice
+// TODO(gbelvin): fully migrate to source slices that encode time directly.
```

Comment:
Are you going to change the wire format of `MapMetadata` proto? If so, this would be going too far in my opinion. Can we stop the rush and discuss it again or maybe do a small doc with options, and technical reasons for doing it? I feel like this started from a small +-1 error problem, and now being inflated to a bunch of readability and complexity problems.